extensions/listening-announcer/src/main/java/io/druid/server/listener/announcer/ListenerAnnouncementResource.java

```diff
@@ -0,0 +1,370 @@
+   *
+   * @param <InObjType>     A List of this type is expected in the input stream as JSON. Must be able to be converted to/from Map<String, Object>
+   * @param <ReturnObjType> The return object can be of any type and will be converted to JSON at the jetty layer.
+   */
+  public static abstract class AbstractAnnouncementPOSTHandler<InObjType, ReturnObjType> implements
+      AnnouncementPOSTHandler
+  {
+    private final TypeReference<? extends InObjType> inObjTypeRef;
+
+    public AbstractAnnouncementPOSTHandler(TypeReference<? extends InObjType> inObjTypeRef)
+    {
+      this.inObjTypeRef = inObjTypeRef;
+    }
+
+    public final Response handle(final InputStream inputStream, final ObjectMapper mapper)
+    {
+      final List<InObjType> inObjList;
+      try {
+        inObjList =
+            Lists.newArrayList(
+                Iterables.transform(
+                    mapper.<List<Map<String, Object>>>readValue(
+                        inputStream, new TypeReference<List<Map<String, Object>>>()
+                        {
+                        }
+                    ),
+                    new Function<Map<String, Object>, InObjType>()
+                    {
+                      @Nullable
+                      @Override
+                      public InObjType apply(Map<String, Object> input)
+                      {
+                        return mapper.convertValue(input, inObjTypeRef);
+                      }
+                    }
+                )
+            );
+      }
+      catch (final IllegalArgumentException iae) {
+        return Response.status(Response.Status.BAD_REQUEST).entity(sanitizeException(iae)).build();
+      }
+      catch (final IOException e) {
+        return Response.status(Response.Status.BAD_REQUEST).entity(sanitizeException(e)).build();
+      }
+      final ReturnObjType returnObj;
+      try {
+        returnObj = handle(inObjList);
+      }
+      catch (Exception e) {
+        log.error(e, "Error handling request");
+        return Response.serverError().entity(sanitizeException(e)).build();
+      }
+      return Response.status(Response.Status.ACCEPTED).entity(returnObj).build();
+    }
+
+    public abstract ReturnObjType handle(List<InObjType> inputObject) throws Exception;
+  }
+
+  public interface AnnouncementIDHandler
+  {
+    Response handle(String id);
+  }
+
+  @Target({ElementType.FIELD, ElementType.PARAMETER, ElementType.METHOD})
+  @Retention(RetentionPolicy.RUNTIME)
+  @BindingAnnotation
+  public @interface POSTHandler
+  {
+  }
+
+  @Target({ElementType.FIELD, ElementType.PARAMETER, ElementType.METHOD})
+  @Retention(RetentionPolicy.RUNTIME)
+  @BindingAnnotation
+  public @interface GETHandler
+  {
+  }
+
+  @Target({ElementType.FIELD, ElementType.PARAMETER, ElementType.METHOD})
+  @Retention(RetentionPolicy.RUNTIME)
+  @BindingAnnotation
+  public @interface DELETEHandler
+  {
+  }
+
+  private static Map<String, Object> sanitizeException(Throwable t)
```

Comment:
move this to some more generic Util, I expect this to be used in multiple Resource classes.